### PR TITLE
Generic rewrite

### PR DESCRIPTION
**Pull request by Ploppz**
### Generic rewrite
The biggest change is that I make everything generic in (the new definition of) `Genome`.
Now, `Genome` is a trait, and `NeuralNetwork` (former `Genome`) is an implementor.
The interface hasn't changed much -- the different is just that you now might have to specify that you want to create a `Population<NeuralNetwork>` or impl an `Environment<NeuralNetwork>`.
The motivation behind this change is to open possibilities to define agents that are composed of several neural networks. It also makes it possible to add alternative neural network implementations.

An implementor of `Genome` is conceptually an "organism" that _"has a genome"_.

What should still be done to finish this part:
* Add examples of composed agents
* Add `fn get_network_genomes(&self) -> Vec<NeuralNetwork>` to `Genome` (this is required to make the telemetry feature work)
* There is a problem how to calculate the distance between two `Genome`s if they have several networks (and how to specify the threshold above which two organisms are considered a "match"). But I think this can be solved by using the `get_network_genomes()` mentioned above.

### Start with `i+o` neurons
It's possible to make organisms of a population start with a certain number of neurons corresponding to the amount of inputs and outputs. Right now that is done like this:

```
let start_genome = NeuralNetwork::with_input_and_output(n_inputs, n_outputs);
let mut population = Population::create_population_from(start_genome, 200);
```

However here's an idea for the future: Make this the default behaviour of `create_population()` - this can be made possible by adding `n_inputs() -> usize` and `n_outputs() -> usize` to the `Environment` trait.

### Fixed a problem with Ctrnn
However, it is still unable to solve the XOR problem. I will try to add arbitrary biases next, and see if that will help.

### ... and some other smaller changes

### CODE REVIEW

**Comment by TLmaK0** [comment]
Thanks @Ploppz, I appreciate a lot your help on this project!!! I have a little question, Do you have an example where the generics for Environment and Population will be used? I mean, of course there will be different implementations of NeuralNetwork but at the end they will be always NeuralNetwork, right? Why not remove generics here and always expect a NeuralNetwork?

**Comment by Ploppz** [comment]
Good question, I am not sure if I have a satisfactory answer to it but I will try to explain the reason why I did this. The key motivation is the possibilities it opens -- possibilities which I might or might not use in my work -- compared to the not-very-invasive changes that had to be done.

The idea I had is that one can use several `NeuralNetwork`s in an organism that work together.
For example:
```
struct Agent {nn1: NeuralNetwork, nn2: NeuralNetwork}
```
or
```
struct Agent {networks: Vec<NeuralNetwork>}
```

Implementation of `Genome` on `Agent` should be trivial.

For my own experiments, I foresee that I will appreciate the possibility of doing this, although I have no concrete idea right now.
The motivation stems from the idea that the human brain has different specialized parts that handle e.g. vision, motor, speech, ... etc. (this is true to some extent)

In this library, these different neural networks could be connected to each other, or they could each compute some separate results that are used in some AI algorithm. There could be a network that predicts the "intrinsic motivation" of a reward-seeking agent to name one example. One could also add hyper-parameters to the genome: for example in the case of a reward-seeking agent with epsilon-greedy exploration, we take a random action with probability `epsilon`, which can now be part of the genome to be evolved.

Another way to think of it, is to "set in stone" part of the organism. Like a designer, you can sculpt the general pathways of information. Putting this limited, sort of 'a priori knowledge', into the organism up-front could be a way to boost evolution (converge faster to the desired organism).

What do you think?

**Comment by TLmaK0** [comment]
I love the idea of connect several neural networks. In this case, Why not simply expect an "agent" instead of "neural network"? so if always we will use an "agent" implementation, why we need to create a generic if always we will work with "agent"?. What I'm afraid is to make the api more and more complicated. For me it's very important that any user can use it although it doesn't know anything about Neural Networks. Only create a method "test" that executes in user environment and call a global loop. It could be we can create something like NeatEnvironment that implements Environment<NeuralNetwork> but I'm not sure that making all this stuff will improve the readability and open the library to extension more than make it without generics. Of course, If you found that this can't be done without generics we can discuss to include it. Also could it be I'm not understanding correctly what you are trying to do :).

**Comment by Ploppz** [comment]
I understand your concern with the complexity of the API, but I still think the generic way is the best way.
Before I go on, let me say that I can revert the 'genericity' in this PR if you still want that. I will try to make a few more points in favor of genericity anyway:

* Regarding the API: take a look at the [example](https://github.com/Ploppz/rustneat/blob/generic_rewrite/examples/simple_sample.rs#L11), lines 11 and 12: very small changes in the API when you just want to use a single neural network.
* If we were to make some `struct Agent` in `rustneat` with, I suppose, an internal `Vec<NeuralNetwork>`, then it means more complex code in `rustneat`, having to evolve several neural networks, and also fixing the way these NNs are evolved. I think it's better to keep the core algorithm simple, yet letting any programmer define their own evolvable composed agent that implements `Genome` -- some extra work for the programmer, if he really wants/needs something else, but everything is possible.
* Consider a scientist that would like to explore the biologically-inspired idea of neuromodulators (dopamine, serotonin, ...). Also consider a scientist that wants to make an agent that learns during a test/episode. Genericity could makes both these scenarios possible (if back propagation is possible/feasible in Ctrnn, not sure). (note: in the former case, he would need to write his own activation function, and probably evolve extra parameters that have to do with the neuromodulators)
* Imo, having `NeuralNetwork` impl `Genome` even makes it a bit cleaner because you can easily see which functions are used by the evolution algorithm, and which functions can be used by the programmer for _using/activating_ the NN.

I personally have no use of the genericity at this very moment, I just thought it makes sense.
If you are not convinced, I will revert the genericity (some time this week).

**Comment by TLmaK0** [comment]
Ok, what if we rename Environment to EnvironmentGeneric and create an Environment that implements EnvironmentGeneric<NeuralNetwork> so that way we don't change the api? What do you think? Then we can have the two options.

**Comment by Ploppz** [comment]
You mean something like `trait Environment = EnvironmentGeneric<NeuralNetwork>`? Doesn't seem possible quite yet: https://github.com/rust-lang/rust/issues/41517

But maybe better, I just learned that default type parameters are possible! So we could have `trait Environment<G = NeuralNetwork>` and `struct Organism<G = NeuralNetwork>`. What do you think?

**Comment by Ploppz** [comment]
Yes, the interface becomes exactly as before https://github.com/Ploppz/rustneat/blob/456702e656c1a4f6d8b959a87c8ef3a2d68b5a64/examples/simple_sample.rs#L11-L12

**Comment by TLmaK0** [comment]
Perfect!!! Thanks @Ploppz for your help on this.

**Comment by Ploppz** [comment]
I will open a new PR when I am more or less finished making changes!
